```
00:53:37.045 ******************** TEST 'Swift(macosx-x86_64) :: Driver/environment-mac.swift' FAILED ********************
00:53:37.045 Script:
00:53:37.045 --
00:53:37.045 : 'RUN: at line 3';   env SDKROOT= '/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swift-macosx-x86_64/bin/swift' -module-cache-path '/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache' -swift-version 4  -driver-use-frontend-path /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swift/test/Driver/Inputs/print-var.sh /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swift/test/Driver/environment-mac.swift DYLD_LIBRARY_PATH | '/usr/bin/python' '/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/swift-macosx-x86_64' --sanitize SOURCE_DIR='/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swift' --use-filecheck '/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/buildbot_incremental/llvm-macosx-x86_64/bin/FileCheck' /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swift/test/Driver/environment-mac.swift
00:53:37.045 --
00:53:37.045 Exit Code: 1
00:53:37.045 
00:53:37.045 Command Output (stderr):
00:53:37.045 --
00:53:37.045 
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-swift-5.1-branch/swift/test/Driver/environment-mac.swift:5:11: error: CHECK: expected string not found in input
00:53:37.045 // CHECK: {{^/usr/lib/swift:}}
00:53:37.045           ^
00:53:37.045 <stdin>:2:1: note: scanning from here
00:53:37.045 
00:53:37.045 ^
00:53:37.045 
00:53:37.045 --
00:53:37.045 
00:53:37.045 ********************
```

--- a/include/swift/AST/SearchPathOptions.h
+++ b/include/swift/AST/SearchPathOptions.h
@@ -72,6 +72,9 @@ public:
 
   /// Don't look in for compiler-provided modules.
   bool SkipRuntimeLibraryImportPaths = false;
+  
+  /// Whether the runtime library path is set to the compiler-relative default.
+  bool RuntimeLibraryPathIsDefault = true;
 
   /// Return a hash code of any components from these options that should
   /// contribute to a Swift Bridging PCH hash.

--- a/include/swift/Frontend/Frontend.h
+++ b/include/swift/Frontend/Frontend.h
@@ -182,7 +182,7 @@ public:
 
   void setMainExecutablePath(StringRef Path);
 
-  void setRuntimeResourcePath(StringRef Path);
+  void setRuntimeResourcePath(StringRef Path, bool IsDefault = false);
 
   void setSDKPath(const std::string &Path);
 

--- a/lib/Frontend/CompilerInvocation.cpp
+++ b/lib/Frontend/CompilerInvocation.cpp
@@ -39,7 +39,7 @@ void CompilerInvocation::setMainExecutablePath(StringRef Path) {
   llvm::sys::path::remove_filename(LibPath); // Remove /swift
   llvm::sys::path::remove_filename(LibPath); // Remove /bin
   llvm::sys::path::append(LibPath, "lib", "swift");
-  setRuntimeResourcePath(LibPath.str());
+  setRuntimeResourcePath(LibPath.str(), /*IsDefault=*/true);
 }
 
 static void updateRuntimeLibraryPaths(SearchPathOptions &SearchPathOpts,
@@ -68,9 +68,11 @@ static void updateRuntimeLibraryPaths(SearchPathOptions &SearchPathOpts,
   }
 }
 
-void CompilerInvocation::setRuntimeResourcePath(StringRef Path) {
+void CompilerInvocation::setRuntimeResourcePath(StringRef Path,
+                                                bool IsDefault) {
   SearchPathOpts.RuntimeResourcePath = Path;
   updateRuntimeLibraryPaths(SearchPathOpts, LangOpts.Target);
+  SearchPathOpts.RuntimeLibraryPathIsDefault = IsDefault;
 }
 
 void CompilerInvocation::setTargetTriple(StringRef Triple) {

--- a/lib/Immediate/Immediate.cpp
+++ b/lib/Immediate/Immediate.cpp
@@ -51,10 +51,22 @@
 using namespace swift;
 using namespace swift::immediate;
 
+/// The path for Swift libraries in the OS.
+#define OS_LIBRARY_PATH "/usr/lib/swift"
+
 static void *loadRuntimeLib(StringRef runtimeLibPathWithName) {
 #if defined(_WIN32)
   return LoadLibraryA(runtimeLibPathWithName.str().c_str());
 #else
+#if defined(__APPLE__) && defined(__MACH__)
+  if (!llvm::sys::path::is_absolute(runtimeLibPathWithName)) {
+    // Try an absolute path search for Swift in the OS first.
+    llvm::SmallString<128> absolutePath(OS_LIBRARY_PATH);
+    llvm::sys::path::append(absolutePath, runtimeLibPathWithName);
+    auto result = dlopen(absolutePath.c_str(), RTLD_LAZY | RTLD_GLOBAL);
+    if (result) return result;
+  }
+#endif
   return dlopen(runtimeLibPathWithName.str().c_str(), RTLD_LAZY | RTLD_GLOBAL);
 #endif
 }
@@ -66,8 +78,16 @@ static void *loadRuntimeLib(StringRef sharedLibName, StringRef runtimeLibPath) {
   return loadRuntimeLib(Path);
 }
 
-void *swift::immediate::loadSwiftRuntime(StringRef runtimeLibPath) {
-  return loadRuntimeLib("libswiftCore" LTDL_SHLIB_EXT, runtimeLibPath);
+void *swift::immediate::loadSwiftRuntime(StringRef runtimeLibPath,
+                                         bool IsDefault) {
+  StringRef LibName = "libswiftCore" LTDL_SHLIB_EXT;
+#if defined(__APPLE__) && defined(__MACH__)
+  if (IsDefault) {
+    auto result = loadRuntimeLib(LibName);
+    if (result) return result;
+  }
+#endif
+  return loadRuntimeLib(LibName, runtimeLibPath);
 }
 
 static bool tryLoadLibrary(LinkLibrary linkLib,
@@ -240,7 +260,9 @@ int swift::RunImmediately(CompilerInstance &CI, const ProcessCmdLine &CmdLine,
   //
   // This must be done here, before any library loading has been done, to avoid
   // racing with the static initializers in user code.
-  auto stdlib = loadSwiftRuntime(Context.SearchPathOpts.RuntimeLibraryPath);
+  auto stdlib = loadSwiftRuntime(
+    Context.SearchPathOpts.RuntimeLibraryPath,
+    Context.SearchPathOpts.RuntimeLibraryPathIsDefault);
   if (!stdlib) {
     CI.getDiags().diagnose(SourceLoc(),
                            diag::error_immediate_mode_missing_stdlib);

--- a/lib/Immediate/ImmediateImpl.h
+++ b/lib/Immediate/ImmediateImpl.h
@@ -38,7 +38,8 @@ namespace immediate {
 /// calls or \c null if an error occurred.
 ///
 /// \param runtimeLibPath Path to search for compiler-relative stdlib dylibs.
-void *loadSwiftRuntime(StringRef runtimeLibPath);
+/// \param IsDefault If true, the path is the default compiler-relative path.
+void *loadSwiftRuntime(StringRef runtimeLibPath, bool IsDefault);
 bool tryLoadLibraries(ArrayRef<LinkLibrary> LinkLibraries,
                       SearchPathOptions SearchPathOpts,
                       DiagnosticEngine &Diags);

--- a/lib/Immediate/REPL.cpp
+++ b/lib/Immediate/REPL.cpp
@@ -969,7 +969,8 @@ public:
     ASTContext &Ctx = CI.getASTContext();
     Ctx.LangOpts.EnableAccessControl = false;
     if (!ParseStdlib) {
-      if (!loadSwiftRuntime(Ctx.SearchPathOpts.RuntimeLibraryPath)) {
+      if (!loadSwiftRuntime(Ctx.SearchPathOpts.RuntimeLibraryPath,
+                            Ctx.SearchPathOpts.RuntimeLibraryPathIsDefault)) {
         CI.getDiags().diagnose(SourceLoc(),
                                diag::error_immediate_mode_missing_stdlib);
         return;

--- a/test/Driver/environment-mac.swift
+++ b/test/Driver/environment-mac.swift
@@ -1,5 +1,5 @@
 // REQUIRES: OS=macosx
 
-// RUN: %swift_driver -driver-use-frontend-path %S/Inputs/print-var.sh %s DYLD_LIBRARY_PATH | %FileCheck %s
+// RUN: %swift_driver -driver-use-frontend-path %S/Inputs/print-var.sh %s DYLD_LIBRARY_PATH | %FileCheck %s --dump-input fail
 
 // CHECK: {{^/usr/lib/swift:}}

--- a/test/Driver/options-interpreter.swift
+++ b/test/Driver/options-interpreter.swift
@@ -15,20 +15,20 @@
 
 
 // RUN: %swift_driver -### -target x86_64-apple-macosx10.9 -resource-dir /RSRC/ %s | %FileCheck -check-prefix=CHECK-RESOURCE-DIR-ONLY %s
-// CHECK-RESOURCE-DIR-ONLY: # DYLD_LIBRARY_PATH=/RSRC/macosx{{$}}
+// CHECK-RESOURCE-DIR-ONLY: # DYLD_LIBRARY_PATH={{(/usr/lib/swift:)?}}/RSRC/macosx{{$}}
 
 // RUN: %swift_driver -### -target x86_64-unknown-linux-gnu -resource-dir /RSRC/ %s | %FileCheck -check-prefix=CHECK-RESOURCE-DIR-ONLY-LINUX${LD_LIBRARY_PATH+_LAX} %s
 // CHECK-RESOURCE-DIR-ONLY-LINUX: # LD_LIBRARY_PATH=/RSRC/linux{{$}}
 // CHECK-RESOURCE-DIR-ONLY-LINUX_LAX: # LD_LIBRARY_PATH=/RSRC/linux{{$|:}}
 
 // RUN: %swift_driver -### -target x86_64-apple-macosx10.9 -L/foo/ %s | %FileCheck -check-prefix=CHECK-L %s
-// CHECK-L: # DYLD_LIBRARY_PATH={{/foo/:[^:]+/lib/swift/macosx$}}
+// CHECK-L: # DYLD_LIBRARY_PATH={{/foo/:(/usr/lib/swift:)?[^:]+/lib/swift/macosx$}}
 
 // RUN: %swift_driver -### -target x86_64-apple-macosx10.9 -L/foo/ -L/bar/ %s | %FileCheck -check-prefix=CHECK-L2 %s
-// CHECK-L2: # DYLD_LIBRARY_PATH={{/foo/:/bar/:[^:]+/lib/swift/macosx$}}
+// CHECK-L2: # DYLD_LIBRARY_PATH={{/foo/:/bar/:(/usr/lib/swift:)?[^:]+/lib/swift/macosx$}}
 
 // RUN: env DYLD_LIBRARY_PATH=/abc/ %swift_driver_plain -### -target x86_64-apple-macosx10.9 -L/foo/ -L/bar/ %s | %FileCheck -check-prefix=CHECK-L2-ENV %s
-// CHECK-L2-ENV: # DYLD_LIBRARY_PATH={{/foo/:/bar/:[^:]+/lib/swift/macosx:/abc/$}}
+// CHECK-L2-ENV: # DYLD_LIBRARY_PATH={{/foo/:/bar/:(/usr/lib/swift:)?[^:]+/lib/swift/macosx:/abc/$}}
 
 // RUN: %swift_driver -### -target x86_64-apple-macosx10.9 %s | %FileCheck -check-prefix=CHECK-NO-FRAMEWORKS %s
 // RUN: env DYLD_FRAMEWORK_PATH=/abc/ %swift_driver_plain -### -target x86_64-apple-macosx10.9 %s | %FileCheck -check-prefix=CHECK-NO-FRAMEWORKS %s
@@ -56,7 +56,7 @@
 // CHECK-COMPLEX: -F /bar/
 // CHECK-COMPLEX: #
 // CHECK-COMPLEX-DAG: DYLD_FRAMEWORK_PATH=/foo/:/bar/:/abc/{{$| }}
-// CHECK-COMPLEX-DAG: DYLD_LIBRARY_PATH={{/foo2/:/bar2/:[^:]+/lib/swift/macosx($| )}}
+// CHECK-COMPLEX-DAG: DYLD_LIBRARY_PATH={{/foo2/:/bar2/:(/usr/lib/swift:)?[^:]+/lib/swift/macosx($| )}}
 
 // RUN: %swift_driver -### -target x86_64-unknown-linux-gnu -L/foo/ %s | %FileCheck -check-prefix=CHECK-L-LINUX${LD_LIBRARY_PATH+_LAX} %s
 // CHECK-L-LINUX: # LD_LIBRARY_PATH={{/foo/:[^:]+/lib/swift/linux$}}

--- a/test/Interpreter/repl_autolinking.swift
+++ b/test/Interpreter/repl_autolinking.swift
@@ -4,7 +4,10 @@
 // RUN: sed -n -e '/REPL_START$/,/REPL_END$/ p' %s > %t/repl.swift
 // RUN: %target-swiftc_driver -emit-library %t/a.swift -I %t -L %t -emit-module-path %t/ModuleA.swiftmodule -autolink-force-load -module-link-name ModuleA -module-name ModuleA -o %t/libModuleA.dylib
 // RUN: %target-swiftc_driver -emit-library %t/b.swift -I %t -L %t -emit-module-path %t/ModuleB.swiftmodule -autolink-force-load -module-link-name ModuleB -module-name ModuleB -o %t/libModuleB.dylib
-// RUN: %swift -repl -I %t -L %t < %t/repl.swift 2>&1 | %FileCheck %s
+// RUN: DYLD_LIBRARY_PATH=/usr/lib/swift %swift -repl -I %t -L %t < %t/repl.swift 2>&1 | %FileCheck %s
+
+// FIXME: remove DYLD_LIBRARY_PATH from the last command when we fix the 
+// interpreter's runtime library lookup.
 
 // REQUIRES: swift_repl
 // UNSUPPORTED: OS=linux-gnu

